openedx_tagging/core/tagging/rest_api/v1/views.py

```diff
@@ -399,6 +405,78 @@ class TaxonomyTagsView(ListAPIView):
         * 400 - Invalid query parameter
         * 403 - Permission denied
         * 404 - Taxonomy not found
+
+    **Create Query Parameters**
+        * pk (required) - The pk of the taxonomy to create a Tag for
+
+    **Create Request Body**
+        * tag (required): The value of the Tag that should be added to
+          the Taxonomy
+        * parent_tag_id (optional): The id of the parent tag that the new
+          Tag should fall under
+        * extenal_id (optional): The external id for the new Tag
+
+    **Create Example Requests**
+        POST api/tagging/v1/taxonomy/:pk/tags                                       - Create a Tag in taxonomy
+        {
+            "value": "New Tag",
+            "parent_tag_id": 123
+            "external_id": "abc123",
+        }
+
+    **Create Query Returns**
+        * 201 - Success
+        * 400 - Invalid parameters provided
+        * 403 - Permission denied
+        * 404 - Taxonomy not found
+
+    **Update Query Parameters**
+        * pk (required) - The pk of the taxonomy to update a Tag in
+
+    **Update Request Body**
+        * tag (required): The ID of the Tag that should be updated
+        * tag_value (required): The updated value of the Tag
+
+    **Update Example Requests**
+        PUT api/tagging/v1/taxonomy/:pk/tags                                        - Update a Tag in Taxonomy
+        {
+            "tag": 1,
+            "tag_value": "Updated Tag Value"
+        }
+        PATCH api/tagging/v1/taxonomy/:pk/tags                                      - Update a Tag in Taxonomy
+        {
+            "tag": 1,
+            "tag_value": "Updated Tag Value"
+        }
+
+    **Update Query Returns**
+        * 200 - Success
+        * 400 - Invalid parameters provided
+        * 403 - Permission denied
+        * 404 - Taxonomy, Tag or Parent Tag not found
+
+    **Delete Query Parameters**
+        * pk (required) - The pk of the taxonomy to Delete Tag(s) in
+
+    **Delete Request Body**
+        * tag_ids (required): The IDs of Tags that should be deleted from Taxonomy
+        * with_subtags (optional): If a Tag in the provided ids contains
+                                   children (subtags), deletion will fail unless
+                                   set to `True`. Defaults to `False`.
+
+    **Delete Example Requests**
+        DELETE api/tagging/v1/taxonomy/:pk/tags                                     - Delete Tag(s) in Taxonomy
+        {
+            "tag_ids": [1,2,3],
+            "with_subtags": True
+        }
+
+    **Delete Query Returns**
+        * 200 - Success
+        * 400 - Invalid parameters provided
+        * 403 - Permission denied
+        * 404 - Taxonomy not found
+
     """
 
     permission_classes = [TagListPermissions]
```

Comment:
@yusuf-musleh I think your best approach would be to modify (and rename) [`TagListPermissions`](https://github.com/openedx/openedx-learning/blob/d44775d83e00e642c96077da93d6ed82c06204be/openedx_tagging/core/tagging/rest_api/v1/permissions.py#L38) to provide a `perm_map` like we did with [`ObjectTagObjectPermissions`](https://github.com/openedx/openedx-learning/blob/d44775d83e00e642c96077da93d6ed82c06204be/openedx_tagging/core/tagging/rest_api/v1/permissions.py#L23).

That will enforce the `oel_tagging.view_tag`, `oel_tagging.change_tag`, and `oel_tagging.delete_tag` rules before invoking the view methods.

But I think you're correct in your question -- these `oel_tagging.*_tag` rules need to be updated to reflect the changes to the taxonomy view/change rules made by https://github.com/openedx/openedx-learning/pull/94.

I think something like this would be sufficient?

```
def can_view_tag(user: UserType, tag: Tag | None = None) -> bool:
    """
    Users can view tags for any taxonomy they can view.
    """"
    taxonomy = tag.taxonomy.cast() if (tag and tag.taxonomy) else None
    return  user.has_perm(
        "oel_tagging.view_taxonomy",
        taxonomy,
    )

def can_change_tag(user: UserType, tag: Tag | None = None) -> bool:
    """
    Users can change tags for any taxonomy they can modify.
    """"
    taxonomy = tag.taxonomy.cast() if (tag and tag.taxonomy) else None
    return  user.has_perm(
        "oel_tagging.change_taxonomy",
        taxonomy,
    )

rules.add_perm("oel_tagging.add_tag", can_change_tag)
rules.add_perm("oel_tagging.change_tag", can_change_tag)
rules.add_perm("oel_tagging.delete_tag", can_change_tag)
rules.add_perm("oel_tagging.view_tag", can_view_tag)
rules.add_perm("oel_tagging.list_tag", can_view_tag)
```

Reply:
@pomegranited Thanks for the pointers! I went ahead with what you suggested, however I faced a few issues and took me quite some time to figure them out:
1. The `get_queryset` method in `TaxonomyTagsView` was initially returning a list, as it does some custom population of an attribute `subtags` and later serialized using the `TagsForSearchSerializer`. For the updated permissions (using `perm_map`) to work,`get_queryset` needs to return a queryset as it is used internally by DRF. The current solution I have (though not a huge fan of it) is getting the result list and constructing a queryset using `filter` and then repopulating the `subtags` again if applicable. Ideally, I think the approach should be to refactor the underlying implementation of [get_matching_tags](https://github.com/openedx/openedx-learning/blob/ca6bc859f69d5508d5457f42a61e8fd3263a7b53/openedx_tagging/core/tagging/rest_api/v1/views.py#L468) to use querysets instead. I couldn't get to that today, but I thought I'd mention it here to get your thoughts.
2. After adding the `perms_map` for `TagObjectPermissions`, I left the `has_object_permission` method in place, since it's a special case for the GET list endpoint. There is not `tag` object in that case, we have the Taxonomy object, so we utilize the `can_view_taxonomy` permission for the `oel_tagging.list_tag` case. It gets [called](https://github.com/openedx/openedx-learning/blob/ca6bc859f69d5508d5457f42a61e8fd3263a7b53/openedx_tagging/core/tagging/rest_api/v1/views.py#L429) in the `get_taxonomy` method to and handle whether the use can view the Tags under this Taxonomy or not.

Reply:
@yusuf-musleh 

> 1... For the updated permissions (using `perm_map`) to work, `get_queryset` needs to return a queryset as it is used internally by DRF... Ideally, I think the approach should be to refactor the underlying implementation of [get_matching_tags](https://github.com/openedx/openedx-learning/blob/ca6bc859f69d5508d5457f42a61e8fd3263a7b53/openedx_tagging/core/tagging/rest_api/v1/views.py#L468) to use querysets instead.

I agree.. have you seen @bradenmacdonald 's  https://github.com/openedx/openedx-learning/pull/92 ? I think that PR does what you need already. We ran out of budget to complete it, but if it's needed for this work, then maybe we can complete it here?

And what you did for the permissions looks great, thank you.

Reply:
@pomegranited I went through @bradenmacdonald 's https://github.com/openedx/openedx-learning/pull/92, lots of great work there, and already does alot of the heavy lifting for the refactoring, however since there are substantial changes made in that PR (and some more work remaining there) I feel it might get pretty messy to include it as part of this PR. Would it make sense to leave the current code in this PR as is and then continue working on https://github.com/openedx/openedx-learning/pull/92 and along with the refactoring needed here as part of a separate ticket? 

Alternatively, I could copy the relevant changes from that https://github.com/openedx/openedx-learning/pull/92 and include it as part of this PR? What do you guys think?

Reply:
> Would it make sense to leave the current code in this PR as is and then continue working on https://github.com/openedx/openedx-learning/pull/92 and along with the refactoring needed here as part of a separate ticket?

Yes, I think that makes sense :)

Reply:
Created https://github.com/openedx/modular-learning/issues/142 for this.

Reply:
Great, thanks for creating it @pomegranited !